presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java

```diff
@@ -193,18 +200,41 @@ public Page getOutput()
         }
 
         if (nestedLoopPageBuilder != null && nestedLoopPageBuilder.hasNext()) {
-            return nestedLoopPageBuilder.next();
+            return applyFilter(nestedLoopPageBuilder.next());
         }
 
         if (buildPageIterator.hasNext()) {
             nestedLoopPageBuilder = new NestedLoopPageBuilder(probePage, buildPageIterator.next());
-            return nestedLoopPageBuilder.next();
+            return applyFilter(nestedLoopPageBuilder.next());
         }
 
         probePage = null;
         return null;
     }
 
+    private Page applyFilter(Page page)
+    {
+        if (!pageFilter.isPresent()) {
+            return page;
+        }
+
+        SelectedPositions selectedPositions = pageFilter.get().filter(operatorContext.getSession().toConnectorSession(), pageFilter.get().getInputChannels().getInputChannels(page));
+        Block[] blocks = page.getBlocks();
+        Block[] filteredBlocks = new Block[blocks.length];
+        if (selectedPositions.isList()) {
+            for (int i = 0; i < blocks.length; i++) {
+                filteredBlocks[i] = blocks[i].copyPositions(selectedPositions.getPositions(), selectedPositions.getOffset(), selectedPositions.size());
```

Comment:
@sopel39 I'm going to extend `getPositions(int[])` to `getPositions(int[], int, int)` like `copyPositions(int[], int, int)`. Do you know if there is anyone has already been working on this? If not, I will just write one.

Reply:
I wonder why you didn't use block masking here? e.g.

```
int[] retainedPositions = Arrays.copyOfRange(
        selectedPositions.getPositions(),
        selectedPositions.getOffset(),
        selectedPositions.getOffset() + selectedPositions.size());
Page filteredPage = page.getPositions(retainedPositions);
```

Reply:
@highker not that I am aware of